src/EventListener/ShippingExportEventListener.php

```diff
@@ -1,65 +1,143 @@
+        /** @var ShippingExportInterface $shippingExport */
+        $shippingExport = $event->getSubject();
+        Assert::isInstanceOf($shippingExport, ShippingExportInterface::class);
+
+        /** @var ShippingGatewayInterface $shippingGateway */
         $shippingGateway = $shippingExport->getShippingGateway();
+        Assert::notNull($shippingGateway);
 
-        if ($shippingGateway->getCode() !== self::POCZTA_POLSKA_GATEWAY_CODE) {
+        $shipment = $shippingExport->getShipment();
+        Assert::notNull($shipment);
+
+        if (self::POCZTA_POLSKA_GATEWAY_CODE !== $shippingGateway->getCode()) {
             return;
         }
 
-        $shipment = $shippingExport->getShipment();
+        try {
+            $this->webClient->setShippingGateway($shippingGateway);
+            $this->webClient->setShipment($shipment);
 
-        $this->webClient->setShippingGateway($shippingGateway);
-        $this->webClient->setShipment($shipment);
+            $this->response = $this->webClient->createLabel();
 
-        try {
-            $label = $this->webClient->createLabel();
-        } catch (\Exception $exception) {
-            $exportShipmentEvent->addErrorFlash(sprintf(
-                "Poczta polska Web Service for #%s order: %s",
-                $shipment->getOrder()->getNumber(),
-                $exception->getMessage()));
+            $labelContent = $this->webClient->getLabelContent();
+            Assert::notNull($labelContent);
+
+            $this->saveShippingLabel($shippingExport, $labelContent, 'pdf');
+        } catch (SoapFault $exception) {
+            $this->flashBag->add(
+                'error',
+                sprintf(
+                    'Poczta Polska Web Service for #%s order: %s',
+                    $shipment->getOrder()->getNumber(),
+                    $exception->getMessage()
+                )
+            );
 
             return;
         }
-//      TODO: Add setTracking to ShippingExportPlugin
-//      $shipment->setTracking($label->content->nrNadania);
 
-        $exportShipmentEvent->saveShippingLabel($label->content->pdfContent, 'pdf');
-        $exportShipmentEvent->addSuccessFlash();
-        $exportShipmentEvent->exportShipment();
+
+        $this->flashBag->add('success', 'bitbag.ui.shipment_data_has_been_exported');
+        $this->markShipmentAsExported($shippingExport);
+    }
+
+    public function saveShippingLabel(
+        ShippingExportInterface $shippingExport,
+        string $labelContent,
+        string $labelExtension
+    ): void {
+        $labelPath = $this->shippingLabelsPath
+            . '/' . $this->getFilename($shippingExport)
+            . '.' . $labelExtension;
+
+        $this->filesystem->dumpFile($labelPath, $labelContent);
+        $shippingExport->setLabelPath($labelPath);
+
+        $this->shippingExportRepository->add($shippingExport);
+    }
+
+    private function getFilename(ShippingExportInterface $shippingExport): string
```

Comment:
Can you extract it into new class and spec it?

Reply:
Function extracted to new ShippingExportFileProvider class and written spec tests